File: sdk/reproit-linux/reproit_linux/signature.py
"""Canonical structural screen signature for native Linux GUI apps.

Byte-identical to the Rust oracle (crates/reproit/src/model/signature.rs), the
React Native SDK (sdk/reproit-react-native/src/signature.ts), the web SDK, and
the Linux runner (runners/linux-atspi.py). Spec: docs/signature.md. Proven
against the 24 golden vectors in signature_vectors.json by
tests/test_parity.py.

A signature hashes STRUCTURE (roles + ids + types + icons + tree shape), never
localized text, so an English and a Japanese render of the same screen hash
identically. The descriptor is:

    "A:" + anchor + "\\n" + tokens.join(";") + value_section

where each retained node emits one pre-order token:

    <depth>:<role>[:<type>][#<icon>][@<id>]   (plus "*" when collapsed)

hashed with FNV-1a 32-bit (offset 0x811c9dc5, prime 0x01000193) into 8
lowercase hex chars.

This module has NO Linux/GTK imports: it is pure Python so the parity test runs
on any host. The live AT-SPI / GTK capture lives in capture.py.
"""

# The fixed, language-independent role vocabulary (docs/signature.md "Roles").
# Anything outside this set normalizes to `node`.
ROLES = frozenset([
    "screen", "header", "text", "button", "link", "textfield", "image", "icon",
    "list", "listitem", "tab", "switch", "checkbox", "radio", "slider", "menu",
    "menuitem", "dialog", "group", "node",
])

# Roles that flicker in and out of the tree and must be dropped before hashing
# (docs/signature.md normalization rule 2). "transient error banner" is not a
# distinct role in the vocabulary, so it is expressed via the `transient` flag on
# a node; both paths drop the node and its whole subtree. `progress` is the role
# name for spinner/progress.
TRANSIENT_ROLES = frozenset([
    "toast", "snackbar", "spinner", "progress", "tooltip", "badge",
])

# Value-role set (docs/signature.md "Value-state", Layer 2). A node is
# value-bearing only if it has a `value` AND its RAW role is one of these (or it
# is explicitly flagged via `value_node`, Layer 3). Chrome roles (button / label
# / header / text / link) are NEVER value-bearing, preserving the rule-1 text
# exclusion. Several of these (status, log, progressbar, meter, timer, output)
# are NOT in the structural ROLES vocabulary, so they normalize to `node` in the
# token body; the value-role test therefore uses the RAW role, not the
# normalized one.
VALUE_ROLES = frozenset([
    "textfield", "status", "log", "progressbar", "meter", "timer", "output",
])


class Node:
    """A normalized accessibility node: the input to the signature.

    Mirrors the Rust `Node` JSON shape so each golden vector's `tree` parses
    directly via Node.from_json:
        { "role": "button", "id": "submit", "type": "text",
          "icon": "e5cd", "transient": false, "value": "3",
          "value_node": false, "children": [ ... ] }
    All fields except `role`/`children` are optional. Localized chrome text is
    excluded from the descriptor by construction (rule 1). `value` is the
    displayed data value (Layer 2) and is consulted ONLY when the node is
    value-bearing (a value-role or `value_node`-flagged); chrome text never goes
    here. Defaults keep a value-less tree byte-identical to a pre-value-state
    one.
    """

    __slots__ = ("role", "id", "type", "icon", "transient", "value", "value_node", "children")

    def __init__(self, role, id=None, type=None, icon=None, transient=False,
                 value=None, value_node=False, children=None):
        self.role = role
        self.id = id
        self.type = type
        self.icon = icon
        self.transient = transient
        self.value = value
        self.value_node = value_node
        self.children = children if children is not None else []

    @staticmethod
    def from_json(j):
        kids = [Node.from_json(c) for c in (j.get("children") or [])]
        return Node(
            role=j["role"],
            id=j.get("id"),
            type=j.get("type"),
            icon=j.get("icon"),
            transient=bool(j.get("transient", False)),
            value=j.get("value"),
            value_node=bool(j.get("value_node", False)),
            children=kids,
        )


def normalize_role(role):
    """Known roles pass through; unknown roles map to `node`."""
    return role if role in ROLES else "node"


def _is_transient(node):
    return node.transient or node.role in TRANSIENT_ROLES


class _NormNode:
    """A node after rules 1, 2, 4 (transients removed, children normalized in
    order). Rule 3 (collapse) is applied at serialization time."""

    __slots__ = ("role", "type", "icon", "id", "children")

    def __init__(self, role, type, icon, id, children):
        self.role = role
        self.type = type
        self.icon = icon
        self.id = id
        self.children = children


def _normalize(node):
    """Apply rules 1, 2, 4: exclude text (no text field exists), drop transient
    subtrees, keep document order. Returns None if this node itself is
    transient."""
    if _is_transient(node):
        return None
    children = []
    for c in node.children:
        nc = _normalize(c)
        if nc is not None:
            children.append(nc)
    return _NormNode(normalize_role(node.role), node.type, node.icon, node.id, children)


def _token_body(n):
    """One node's token body (everything after `<depth>:`), without the repeat
    marker: `<role>[:<type>][#<icon>][@<id>]`."""
    s = n.role
    if n.type is not None:
        s += ":" + n.type
    if n.icon is not None:
        s += "#" + n.icon
    if n.id is not None:
        s += "@" + n.id
    return s


def _subtree_key(n):
    """The canonical subtree descriptor used for collapse comparison (rule 3):
    the pre-order token list of this subtree, depths re-based to 0, so two
    sibling subtrees at the same level compare equal regardless of absolute
    depth."""
    tokens = []
    _walk_key(n, 0, tokens)
    return ";".join(tokens)


def _walk_key(n, depth, tokens):
    tokens.append("%d:%s" % (depth, _token_body(n)))
    for c in n.children:
        _walk_key(c, depth + 1, tokens)


def _serialize_node(n, depth, repeated, tokens):
    """Emit one node's token (optionally marked repeated) then recurse,
    collapsing across the children run."""
    tok = "%d:%s" % (depth, _token_body(n))
    if repeated:
        tok += "*"
    tokens.append(tok)
    _serialize_children(n.children, depth + 1, tokens)


def _serialize_children(children, depth, tokens):
    """Walk a run of siblings, collapsing maximal runs of >= 2 consecutive
    children whose subtree_key is identical into one emission with the `*`
    marker (count dropped)."""
    i = 0
    while i < len(children):
        key = _subtree_key(children[i])
        j = i + 1
        while j < len(children) and _subtree_key(children[j]) == key:
            j += 1
        run = j - i
        _serialize_node(children[i], depth, run >= 2, tokens)
        i = j


# --- Layer 2: value-state (docs/signature.md "Value-state") -----------------

def is_value_bearing(node):
    """True iff this node carries a canonical value-class in the V: section: it
    has a `value` AND it is value-bearing, i.e. its RAW role is a value-role OR
    it is `value_node`-flagged (Layer 3 opt-in). The raw role is used
    deliberately: roles like `status`/`meter` normalize to `node` but are still
    value-roles."""
    return node.value is not None and (node.role in VALUE_ROLES or node.value_node)


def value_class(s):
    r"""Map a value string to a bounded, deterministic, locale-safe value-class
    token. The numeric branch accepts ONLY the strict period-decimal grammar
    `^[+-]?[0-9]+(\.[0-9]+)?$` (no grouping, no exponent, no leading/trailing
    dot, ASCII digits only); anything ambiguous falls back to NONEMPTY because we
    do not guess locale formats."""
    t = s.strip()
    if t == "":
        return "EMPTY"
    if _is_strict_decimal(t):
        n = float(t)
        a = abs(n)
        if n == 0.0:
            return "ZERO"
        if n < 0.0:
            return "NEG"
        if a < 10.0:
            return "POS1"
        if a < 100.0:
            return "POS2"
        if a < 1000.0:
            return "POS3"
        return "POSL"
    return "NONEMPTY"


def _is_strict_decimal(s):
    r"""Strict `^[+-]?[0-9]+(\.[0-9]+)?$`: optional sign, one or more ASCII
    digits, optionally a period plus one or more ASCII digits. Locale-safe by
    construction (no grouping separators, no exponent, no bare dot)."""
    i = 0
    n = len(s)
    if i < n and (s[i] == "+" or s[i] == "-"):
        i += 1
    int_start = i
    while i < n and "0" <= s[i] <= "9":
        i += 1
    if i == int_start:
        return False  # need at least one integer digit
    if i < n and s[i] == ".":
        i += 1
        frac_start = i
        while i < n and "0" <= s[i] <= "9":
            i += 1
        if i == frac_start:
            return False  # a trailing dot with no fraction digits is not allowed
    return i == n


def _value_key(node, structural_index):
    """The V:-section key for a value-bearing node: its stable `id` as `key:<id>`
    if present, else the structural fallback `role:<role>#<idx>` using the
    NORMALIZED role (so the key namespace matches the selector grammar)."""
    if node.id is not None:
        return "key:%s" % node.id
    return "role:%s#%d" % (normalize_role(node.role), structural_index)


def _collect_values(node, out):
    """Collect (value_key, value_class) for the root then descend. Transient
    subtrees (rule 2) are skipped so the V: section stays consistent with the
    body. The root has no peers, so its keyless structural index is 0."""
    if _is_transient(node):
        return
    if is_value_bearing(node):
        out.append((_value_key(node, 0), value_class(node.value if node.value is not None else "")))
    _collect_values_children(node, out)


def _collect_values_children(node, out):
    """Descend into non-transient children, assigning each keyless child its
    per-parent structural index among same-normalized-role peers (matching the
    selector `#idx`), emitting value-bearing children, then recursing. The node
    itself is NOT re-emitted (the caller already handled it)."""
    role_counts = {}
    for child in node.children:
        if _is_transient(child):
            continue
        role = normalize_role(child.role)
        idx = role_counts.get(role, 0)
        role_counts[role] = idx + 1
        if is_value_bearing(child):
            out.append((_value_key(child, idx), value_class(child.value if child.value is not None else "")))
        _collect_values_children(child, out)


def _value_section(root):
    r"""Build the V: section suffix. Returns "" when there are NO value-bearing
    nodes, keeping the descriptor (and hash) byte-identical to a pre-value-state
    tree. Otherwise `"\nV:" + key=class;key=class...` sorted by key."""
    pairs = []
    _collect_values(root, pairs)
    if not pairs:
        return ""
    pairs.sort(key=lambda kc: kc[0])
    body = ";".join("%s=%s" % (k, c) for k, c in pairs)
    return "\nV:%s" % body


def descriptor(anchor, root):
    r"""Build the exact UTF-8 descriptor string that gets hashed
    (docs/signature.md "Descriptor serialization"):
    `"A:" + anchor + "\n" + tokens.join(";") + value_section`. The `A:` prefix
    line is always present, even with no anchor. The V: section (Layer 2
    value-classes) is appended ONLY when at least one value-bearing node exists,
    so a value-less tree is byte-identical to the pre-value-state descriptor."""
    tokens = []
    norm = _normalize(root)
    if norm is not None:
        _serialize_node(norm, 0, False, tokens)
    return "A:%s\n%s%s" % (anchor or "", ";".join(tokens), _value_section(root))


def fnv1a32_hex(data):
    """FNV-1a, 32-bit, over `data` bytes; 8-char zero-padded lowercase hex
    (docs/signature.md "Hash"). Offset basis 0x811c9dc5, prime 0x01000193."""
    h = 0x811C9DC5
    for b in data:
        h ^= b
        h = (h * 0x01000193) & 0xFFFFFFFF
    return "%08x" % h


def signature(anchor, root):
    """THE canonical signature: FNV-1a 32-bit over descriptor(), 8 hex chars."""
    return fnv1a32_hex(descriptor(anchor, root).encode("utf-8"))


def selector_for(id, role, structural_index):
    """`key:<id>` when a stable id exists, else `role:<role>#<idx>`. The second
    return value (nokey) is metadata for `map show`; it does NOT affect the
    hash."""
    if id is not None:
        return ("key:%s" % id, False)
    return ("role:%s#%d" % (normalize_role(role), structural_index), True)

File: sdk/reproit-windows/src/ReproIt.Core/Signature.cs
// Canonical structural screen signature for .NET (WPF / WinUI 3).
//
// This is the C# port of the Rust parity oracle
// (crates/reproit/src/model/signature.rs). docs/signature.md is the spec;
// signature_vectors.json (repo root) holds the golden vectors every
// implementation must reproduce bit-for-bit. The production SDK (the WPF / WinUI
// capture in ReproIt.Windows) and the host parity test compute the signature
// through THIS file, so they agree by construction.
//
// The descriptor string that gets hashed is built exactly as the spec defines:
//
//   token = "<depth>:<role>[:<type>][#<icon>][@<id>]" (trailing "*" on a repeat)
//   body  = tokens joined by ";", pre-order
//   desc  = "A:" + anchor + "\n" + body  (+ a trailing "\nV:" value section)
//   sig   = FNV-1a 32-bit over the UTF-8 bytes of desc, 8-char lowercase hex
//
// This file targets netstandard2.0 / net8.0 and has NO Windows (WPF / WinUI)
// dependency on purpose: it is plain C# so the parity test runs cross-platform
// on any host with the .NET SDK, exactly like the Kotlin/Swift signature cores
// run on the host JVM / Foundation without the platform SDK.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace ReproIt.Core
{
    /// <summary>
    /// A normalized accessibility node: the input to the signature.
    ///
    /// Mirrors the Rust <c>Node</c> JSON shape so each golden vector's <c>tree</c>
    /// parses directly:
    /// <code>
    /// { "role": "button", "id": "submit", "type": "text",
    ///   "icon": "e5cd", "transient": false, "value": "3",
    ///   "value_node": false, "children": [ ... ] }
    /// </code>
    /// All fields except <c>role</c> and <c>children</c> are optional. There is
    /// deliberately no text/label field: localized text is excluded from the
    /// descriptor by construction (rule 1), so there is nothing to hash. The
    /// <c>value</c> field is the displayed data value (Layer 2) and is consulted
    /// ONLY when the node is value-bearing (a value-role or value_node-flagged);
    /// chrome text never goes here, so a value-less tree is byte-identical to a
    /// pre-value-state tree.
    /// </summary>
    public sealed class Node
    {
        /// <summary>Role from the fixed vocabulary; unknown roles normalize to "node".</summary>
        public string Role { get; set; }

        /// <summary>Stable developer id (AutomationId / x:Name / a11y-id). Null if none.</summary>
        public string Id { get; set; }

        /// <summary>Optional input-type refinement (text, password, email, number, ...).</summary>
        public string Type { get; set; }

        /// <summary>Optional language-independent icon identity (codepoint / symbol / asset).</summary>
        public string Icon { get; set; }

        /// <summary>Explicit transient marker (e.g. a transient error banner). Dropped
        /// like a transient role, together with its whole subtree.</summary>
        public bool Transient { get; set; }

        /// <summary>The node's displayed data value (Layer 2, docs/signature.md
        /// "Value-state"). Only consulted when the node is value-bearing. Null by
        /// default.</summary>
        public string Value { get; set; }

        /// <summary>Opt-in value-node flag (Layer 3). When true the node is treated
        /// as value-bearing even if its role is not in the value-role set.</summary>
        public bool ValueNode { get; set; }

        /// <summary>Ordered children, in document order.</summary>
        public List<Node> Children { get; set; }

        public Node(string role)
        {
            Role = role;
            Children = new List<Node>();
        }

        public Node() : this("node")
        {
        }
    }

    /// <summary>
    /// The canonical structural-signature core. Static, pure, allocation-light.
    /// Every method here mirrors the Rust oracle line-for-line so the descriptor
    /// it builds is byte-identical across all SDKs and runners.
    /// </summary>
    public static class Signature
    {
        /// <summary>The fixed, language-independent role vocabulary (docs/signature.md
        /// "Roles"). Anything outside this set normalizes to "node".</summary>
        public static readonly HashSet<string> Roles = new HashSet<string>(StringComparer.Ordinal)
        {
            "screen", "header", "text", "button", "link", "textfield", "image",
            "icon", "list", "listitem", "tab", "switch", "checkbox", "radio",
            "slider", "menu", "menuitem", "dialog", "group", "node",
        };

        /// <summary>Roles that flicker in and out of the tree and must be dropped
        /// before hashing (docs/signature.md normalization rule 2). "transient error
        /// banner" is not a distinct role in the vocabulary, so it is expressed via
        /// the <see cref="Node.Transient"/> flag; both paths drop the node and its
        /// whole subtree. "progress" is the role name for spinner/progress.</summary>
        public static readonly HashSet<string> TransientRoles = new HashSet<string>(StringComparer.Ordinal)
        {
            "toast", "snackbar", "spinner", "progress", "tooltip", "badge",
        };

        /// <summary>Value-role set (docs/signature.md "Value-state", Layer 2). A node
        /// carries a canonical value-class in the "V:" section only if it has a
        /// <see cref="Node.Value"/> AND either its RAW role is in this set OR it is
        /// flagged <see cref="Node.ValueNode"/> (the Layer 3 opt-in). Several of these
        /// (status, log, progressbar, meter, timer, output) are NOT in the structural
        /// <see cref="Roles"/> vocabulary, so they normalize to "node" in the
        /// descriptor body; the value-role test therefore uses the RAW role, not the
        /// normalized one. Chrome roles (button/header/text/link) are NEVER
        /// value-bearing, so rule 1's chrome-text exclusion is preserved exactly.</summary>
        public static readonly HashSet<string> ValueRoles = new HashSet<string>(StringComparer.Ordinal)
        {
            "textfield", "status", "log", "progressbar", "meter", "timer", "output",
        };

        /// <summary>Normalize a role to the fixed vocabulary: known roles pass
        /// through, unknown roles map to "node" (docs/signature.md "Roles").</summary>
        public static string NormalizeRole(string role)
        {
            return Roles.Contains(role) ? role : "node";
        }

        private static bool IsTransient(Node n)
        {
            return n.Transient || TransientRoles.Contains(n.Role);
        }

        // ---- Layer 2: bounded, locale-safe value-classes ------------------------
        // (docs/signature.md "Value-state").

        /// <summary>True if <paramref name="n"/> carries a canonical value-class in
        /// the "V:" section: it has a <see cref="Node.Value"/> AND it is value-bearing,
        /// i.e. its RAW role is a value-role OR it is value_node-flagged. The raw role
        /// is used deliberately (roles like "status"/"meter" normalize to "node" but
        /// are still value-roles). Mirrors the oracle's is_value_bearing exactly.</summary>
        private static bool IsValueBearing(Node n)
        {
            return n.Value != null && (ValueRoles.Contains(n.Role) || n.ValueNode);
        }

        /// <summary>Strict <c>^[+-]?[0-9]+(\.[0-9]+)?$</c>: an optional sign, one or
        /// more ASCII digits, optionally a period followed by one or more ASCII digits.
        /// No grouping separators, no exponent, no leading/trailing dot. Locale-safe by
        /// construction. Mirrors the oracle's is_strict_decimal byte-for-byte (it scans
        /// raw chars, never CultureInfo-aware parsing).</summary>
        private static bool IsStrictDecimal(string s)
        {
            int i = 0;
            int n = s.Length;
            if (i < n && (s[i] == '+' || s[i] == '-'))
            {
                i++;
            }
            int intStart = i;
            while (i < n && s[i] >= '0' && s[i] <= '9')
            {
                i++;
            }
            if (i == intStart)
            {
                return false; // need at least one integer digit
            }
            if (i < n && s[i] == '.')
            {
                i++;
                int fracStart = i;
                while (i < n && s[i] >= '0' && s[i] <= '9')
                {
                    i++;
                }
                if (i == fracStart)
                {
                    return false; // a trailing dot with no fraction digits is not allowed
                }
            }
            return i == n;
        }

        /// <summary>Map a value string to a bounded, deterministic, locale-safe
        /// value-class token (docs/signature.md "Value-state"): EMPTY / ZERO / NEG /
        /// POS1 (&lt;10) / POS2 (&lt;100) / POS3 (&lt;1000) / POSL (&gt;=1000) for the
        /// strict period-decimal grammar; NONEMPTY for anything ambiguous
        /// (grouped/locale numbers, currency, exponent, non-ASCII digits, text)
        /// because we do not guess locale formats. Identical buckets to the oracle.</summary>
        public static string ValueClass(string s)
        {
            string t = (s ?? string.Empty).Trim();
            if (t.Length == 0)
            {
                return "EMPTY";
            }
            if (IsStrictDecimal(t))
            {
                // The grammar is a strict subset of InvariantCulture's accepted
                // double syntax, so this parse never fails; InvariantCulture pins the
                // period as the decimal separator regardless of the host locale.
                double num = double.Parse(t, NumberStyles.AllowLeadingSign | NumberStyles.AllowDecimalPoint, CultureInfo.InvariantCulture);
                double a = Math.Abs(num);
                if (num == 0.0)
                {
                    return "ZERO";
                }
                if (num < 0.0)
                {
                    return "NEG";
                }
                if (a < 10.0)
                {
                    return "POS1";
                }
                if (a < 100.0)
                {
                    return "POS2";
                }
                if (a < 1000.0)
                {
                    return "POS3";
                }
                return "POSL";
            }
            return "NONEMPTY";
        }

        /// <summary>The "V:"-section key for a value-bearing node: its stable id as
        /// "key:&lt;id&gt;" if present, otherwise the structural fallback
        /// "role:&lt;role&gt;#&lt;idx&gt;" using the NORMALIZED role (so the key
        /// namespace matches the selector grammar). This is the "stable-key" the "V:"
        /// section sorts on.</summary>
        private static string ValueKey(Node n, int structuralIndex)
        {
            if (n.Id != null)
            {
                return "key:" + n.Id;
            }
            return "role:" + NormalizeRole(n.Role) + "#" + structuralIndex.ToString(CultureInfo.InvariantCulture);
        }

        /// <summary>Collect (value_key, value_class) pairs for every value-bearing
        /// node in the tree, in pre-order, skipping transient subtrees (rule 2) so the
        /// "V:" section stays consistent with the structural body. The root has no
        /// peers, so it gets index 0; each keyless child gets its position among
        /// same-(normalized-)role, non-transient siblings under the same parent. The
        /// result is later sorted by key. Mirrors collect_values + collect_values_children.</summary>
        private static void CollectValues(Node node, List<KeyValuePair<string, string>> outPairs)
        {
            if (IsTransient(node))
            {
                return;
            }
            if (IsValueBearing(node))
            {
                outPairs.Add(new KeyValuePair<string, string>(ValueKey(node, 0), ValueClass(node.Value)));
            }
            CollectValuesChildren(node, outPairs);
        }

        private static void CollectValuesChildren(Node node, List<KeyValuePair<string, string>> outPairs)
        {
            var roleCounts = new Dictionary<string, int>(StringComparer.Ordinal);
            var children = node.Children ?? EmptyChildren;
            foreach (var child in children)
            {
                if (IsTransient(child))
                {
                    continue;
                }
                string role = NormalizeRole(child.Role);
                int idx;
                roleCounts.TryGetValue(role, out idx);
                roleCounts[role] = idx + 1;
                if (IsValueBearing(child))
                {
                    outPairs.Add(new KeyValuePair<string, string>(ValueKey(child, idx), ValueClass(child.Value)));
                }
                CollectValuesChildren(child, outPairs);
            }
        }

        /// <summary>The "V:" section suffix (docs/signature.md "Value-state"). Empty
        /// string when there are NO value-bearing nodes, which keeps the descriptor
        /// (and hash) byte-identical to a pre-value-state tree (backward-compatible).
        /// Otherwise returns "\nV:" + sorted key=class entries joined by ";".</summary>
        private static string ValueSection(Node root)
        {
            var pairs = new List<KeyValuePair<string, string>>();
            CollectValues(root, pairs);
            if (pairs.Count == 0)
            {
                return string.Empty;
            }
            // Sort by the UTF-8 BYTE sequence of the key, matching the Rust
            // oracle's String::cmp (== Unicode code-POINT order). Note: string
            // .CompareOrdinal is UTF-16 code-UNIT order, which DIVERGES for astral
            // chars (surrogate pairs 0xD800-0xDBFF sort below high-BMP chars under
            // UTF-16, but above them under UTF-8/code-point order).
            pairs.Sort((a, b) => CompareUtf8(a.Key, b.Key));
            var sb = new StringBuilder();
            sb.Append("\nV:");
            for (int i = 0; i < pairs.Count; i++)
            {
                if (i > 0)
                {
                    sb.Append(';');
                }
                sb.Append(pairs[i].Key);
                sb.Append('=');
                sb.Append(pairs[i].Value);
            }
            return sb.ToString();
        }

        /// <summary>Compare two strings by their UTF-8 byte sequences, matching the
        /// Rust oracle's <c>String::cmp</c> (== Unicode code-POINT order). This is
        /// NOT the same as <c>string.CompareOrdinal</c> (UTF-16 code-UNIT order),
        /// which diverges for astral chars (surrogate pairs sort below high-BMP
        /// chars under UTF-16, but above them under UTF-8/code-point order).</summary>
        private static int CompareUtf8(string a, string b)
        {
            byte[] ba = Encoding.UTF8.GetBytes(a);
            byte[] bb = Encoding.UTF8.GetBytes(b);
            int n = ba.Length < bb.Length ? ba.Length : bb.Length;
            for (int i = 0; i < n; i++)
            {
                if (ba[i] != bb[i])
                {
                    return ba[i] - bb[i];
                }
            }
            return ba.Length - bb.Length;
        }

        // ---- structural body (rules 1-4) ----------------------------------------

        /// <summary>A normalized node after rules 1, 2, 4 are applied (transients
        /// removed, children normalized in order). Rule 3 (collapse) is applied at
        /// serialization time over the children of this tree.</summary>
        private sealed class NormNode
        {
            public string Role;
            public string Type;
            public string Icon;
            public string Id;
            public List<NormNode> Children;
        }

        private static readonly List<Node> EmptyChildren = new List<Node>();

        /// <summary>Apply rules 1, 2, 4: exclude text (no text field exists), drop
        /// transient subtrees, keep document order. Returns null if this node itself
        /// is transient (caller drops it).</summary>
        private static NormNode Normalize(Node node)
        {
            if (IsTransient(node))
            {
                return null;
            }
            var children = new List<NormNode>();
            var src = node.Children ?? EmptyChildren;
            foreach (var c in src)
            {
                var nc = Normalize(c);
                if (nc != null)
                {
                    children.Add(nc);
                }
            }
            return new NormNode
            {
                Role = NormalizeRole(node.Role),
                Type = node.Type,
                Icon = node.Icon,
                Id = node.Id,
                Children = children,
            };
        }

        /// <summary>One node's token body (everything after "&lt;depth&gt;:"), without
        /// the repeat marker: "&lt;role&gt;[:&lt;type&gt;][#&lt;icon&gt;][@&lt;id&gt;]".</summary>
        private static string TokenBody(NormNode n)
        {
            var sb = new StringBuilder(n.Role);
            if (n.Type != null)
            {
                sb.Append(':');
                sb.Append(n.Type);
            }
            if (n.Icon != null)
            {
                sb.Append('#');
                sb.Append(n.Icon);
            }
            if (n.Id != null)
            {
                sb.Append('@');
                sb.Append(n.Id);
            }
            return sb.ToString();
        }

        /// <summary>The canonical subtree descriptor used for collapse comparison
        /// (rule 3): the pre-order token list of this subtree, depths re-based to 0, so
        /// two sibling subtrees at the same level compare equal regardless of absolute
        /// depth.</summary>
        private static string SubtreeKey(NormNode n)
        {
            var tokens = new List<string>();
            WalkKey(n, 0, tokens);
            return string.Join(";", tokens);
        }

        private static void WalkKey(NormNode n, int depth, List<string> tokens)
        {
            tokens.Add(depth.ToString(CultureInfo.InvariantCulture) + ":" + TokenBody(n));
            foreach (var c in n.Children)
            {
                WalkKey(c, depth + 1, tokens);
            }
        }

        /// <summary>Emit one node's token (optionally marked repeated) then recurse,
        /// collapsing across the children run.</summary>
        private static void SerializeNode(NormNode n, int depth, bool repeated, List<string> tokens)
        {
            string tok = depth.ToString(CultureInfo.InvariantCulture) + ":" + TokenBody(n);
            if (repeated)
            {
                tok += "*";
            }
            tokens.Add(tok);
            SerializeChildren(n.Children, depth + 1, tokens);
        }

        /// <summary>Walk a run of siblings, collapsing maximal runs of &gt;= 2
        /// consecutive children whose SubtreeKey is identical into a single emission
        /// with the "*" marker (count dropped).</summary>
        private static void SerializeChildren(List<NormNode> children, int depth, List<string> tokens)
        {
            int i = 0;
            while (i < children.Count)
            {
                string key = SubtreeKey(children[i]);
                int j = i + 1;
                while (j < children.Count && SubtreeKey(children[j]) == key)
                {
                    j++;
                }
                int run = j - i;
                SerializeNode(children[i], depth, run >= 2, tokens);
                i = j;
            }
        }

        /// <summary>Build the exact UTF-8 descriptor string that gets hashed
        /// (docs/signature.md "Descriptor serialization"):
        /// "A:" + anchor + "\n" + tokens.join(";"), with the Layer 2 "V:" section
        /// appended only when at least one value-bearing node exists. The "A:" prefix
        /// line is always present, even with no anchor (then it is the empty string
        /// "A:" + newline). A tree with no value-bearing nodes is byte-identical to a
        /// pre-value-state tree.</summary>
        public static string Descriptor(string anchor, Node root)
        {
            var tokens = new List<string>();
            var norm = Normalize(root);
            if (norm != null)
            {
                SerializeNode(norm, 0, false, tokens);
            }
            return "A:" + (anchor ?? string.Empty) + "\n" + string.Join(";", tokens) + ValueSection(root);
        }

        /// <summary>THE canonical structural signature: FNV-1a 32-bit over the UTF-8
        /// bytes of <see cref="Descriptor"/>, 8-char zero-padded lowercase hex
        /// (docs/signature.md "Hash").</summary>
        public static string Of(string anchor, Node root)
        {
            string desc = Descriptor(anchor, root);
            return Fnv1a32Hex(Encoding.UTF8.GetBytes(desc));
        }

        /// <summary>FNV-1a, 32-bit, over <paramref name="bytes"/>; 8-char zero-padded
        /// lowercase hex (docs/signature.md "Hash").
        ///
        /// C# <c>uint</c> is unsigned and wraps on overflow by definition, which is
        /// exactly the 32-bit modular arithmetic FNV-1a needs; we do not even need an
        /// <c>unchecked</c> block for an unsigned multiply. Operating over UTF-8 BYTES
        /// (not chars) keeps non-ASCII descriptors byte-identical to the Rust oracle;
        /// ASCII descriptors (the common case) are unaffected.</summary>
        public static string Fnv1a32Hex(byte[] bytes)
        {
            uint h = 0x811c9dc5u; // FNV-1a 32-bit offset basis
            foreach (byte b in bytes)
            {
                h ^= b;
                h *= 0x01000193u; // FNV prime; unsigned multiply wraps mod 2^32
            }
            return h.ToString("x8", CultureInfo.InvariantCulture);
        }

        // ---- Selectors (docs/signature.md "Selectors") --------------------------

        /// <summary>A selector that addresses an element for actions / repros:
        /// id &gt; type+role &gt; role + structural-index. Returns "key:&lt;id&gt;" if
        /// the node has a stable id; otherwise the structural "role:&lt;role&gt;#&lt;idx&gt;"
        /// form. <see cref="NoKey"/> is true whenever no stable id was available
        /// (metadata for "map show"; it does NOT affect the hash).</summary>
        public struct SelectorResult
        {
            public string Selector;
            public bool NoKey;
        }

        /// <summary>Build a selector for a node given its structural index among peers.</summary>
        public static SelectorResult Selector(Node node, int structuralIndex)
        {
            if (node.Id != null)
            {
                return new SelectorResult { Selector = "key:" + node.Id, NoKey = false };
            }
            return new SelectorResult
            {
                Selector = "role:" + NormalizeRole(node.Role) + "#" + structuralIndex.ToString(CultureInfo.InvariantCulture),
                NoKey = true,
            };
        }
    }
}
